hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/simple/HoodieSimpleIndex.java

```diff
@@ -143,19 +141,17 @@ protected <R> HoodieData<HoodieRecord<R>> tagLocationInternal(
   protected HoodiePairData<HoodieKey, HoodieRecordLocation> fetchRecordLocationsForAffectedPartitions(
       HoodieData<HoodieKey> hoodieKeys, HoodieEngineContext context, HoodieTable hoodieTable,
       int parallelism) {
-    List<String> affectedPartitionPathList =
-        hoodieKeys.map(HoodieKey::getPartitionPath).distinct().collectAsList();
-    List<Pair<String, HoodieBaseFile>> latestBaseFiles =
+    HoodieData<String> affectedPartitionPathList =
+        hoodieKeys.map(HoodieKey::getPartitionPath).distinct();
+    HoodieData<Pair<String, HoodieBaseFile>> latestBaseFiles =
         getLatestBaseFilesForAllPartitions(affectedPartitionPathList, context, hoodieTable);
-    return fetchRecordLocations(context, hoodieTable, parallelism, latestBaseFiles);
+    return fetchRecordLocations(hoodieTable, parallelism, latestBaseFiles);
   }
 
   protected HoodiePairData<HoodieKey, HoodieRecordLocation> fetchRecordLocations(
-      HoodieEngineContext context, HoodieTable hoodieTable, int parallelism,
-      List<Pair<String, HoodieBaseFile>> baseFiles) {
-    int fetchParallelism = Math.max(1, Math.min(baseFiles.size(), parallelism));
-
-    return context.parallelize(baseFiles, fetchParallelism)
+      HoodieTable hoodieTable, int parallelism,
+      HoodieData<Pair<String, HoodieBaseFile>> baseFiles) {
+    return baseFiles.repartition(Math.max(1, Math.min(baseFiles.getNumPartitions(), parallelism)))
```

Comment:
should we consider doing coalesce or repartition based on total partitions.

Reply:
There is no coalesce option on HoodieData

Reply:
we can introduce on a need basis. may be, so far we did not have a need.